java/storage/connectors/redis/src/main/java/feast/storage/connectors/redis/retriever/OnlineRetriever.java

```diff
@@ -46,74 +45,90 @@ public OnlineRetriever(RedisClientAdapter redisClientAdapter, EntityKeySerialize
-
-    return features;
+    return getFeaturesFromRedis(redisKeys, featureReferences);
   }
 
-  private List<List<Feature>> getFeaturesFromRedis(
+  private List<Map<ServingAPIProto.FeatureReferenceV2, Feature>> getFeaturesFromRedis(
       List<RedisProto.RedisKeyV2> redisKeys,
       List<ServingAPIProto.FeatureReferenceV2> featureReferences) {
     List<List<Feature>> features = new ArrayList<>();
     // To decode bytes back to Feature Reference
-    Map<byte[], ServingAPIProto.FeatureReferenceV2> byteToFeatureReferenceMap = new HashMap<>();
+    Map<ByteBuffer, ServingAPIProto.FeatureReferenceV2> byteToFeatureReferenceMap = new HashMap<>();
 
     // Serialize using proto
     List<byte[]> binaryRedisKeys =
         redisKeys.stream().map(this.keySerializer::serialize).collect(Collectors.toList());
 
-    List<byte[]> featureReferenceWithTsByteList = new ArrayList<>();
+    List<byte[]> retrieveFields = new ArrayList<>();
     featureReferences.stream()
         .forEach(
             featureReference -> {
 
               // eg. murmur(<featuretable_name:feature_name>)
               byte[] featureReferenceBytes =
                   RedisHashDecoder.getFeatureReferenceRedisHashKeyBytes(featureReference);
-              featureReferenceWithTsByteList.add(featureReferenceBytes);
-              byteToFeatureReferenceMap.put(featureReferenceBytes, featureReference);
+              retrieveFields.add(featureReferenceBytes);
+              byteToFeatureReferenceMap.put(
+                  ByteBuffer.wrap(featureReferenceBytes), featureReference);
+            });
 
+    featureReferences.stream()
+        .map(ServingAPIProto.FeatureReferenceV2::getFeatureTable)
+        .distinct()
+        .forEach(
+            table -> {
               // eg. <_ts:featuretable_name>
               byte[] featureTableTsBytes =
-                  RedisHashDecoder.getTimestampRedisHashKeyBytes(featureReference, timestampPrefix);
-              featureReferenceWithTsByteList.add(featureTableTsBytes);
+                  RedisHashDecoder.getTimestampRedisHashKeyBytes(table, timestampPrefix);
+
+              retrieveFields.add(featureTableTsBytes);
             });
 
-    // Perform a series of independent calls
-    List<RedisFuture<List<KeyValue<byte[], byte[]>>>> futures = Lists.newArrayList();
-    for (byte[] binaryRedisKey : binaryRedisKeys) {
-      byte[][] featureReferenceWithTsByteArrays =
-          featureReferenceWithTsByteList.toArray(new byte[0][]);
-      // Access redis keys and extract features
-      futures.add(redisClientAdapter.hmget(binaryRedisKey, featureReferenceWithTsByteArrays));
+    List<Future<Map<byte[], byte[]>>> futures =
+        Lists.newArrayListWithExpectedSize(binaryRedisKeys.size());
+
+    // Number of fields that controls whether to use hmget or hgetall was discovered empirically
+    // Could be potentially tuned further
+    if (retrieveFields.size() < 50) {
```

Comment:
nit: maybe make this a constant at the beginning of the file so it's better visible?

Reply:
Added constant